### PR TITLE
Iter13

### PR DESCRIPTION
**Pull request by FoGezz**

### CODE REVIEW

File: cmd/shortener/routes.go
Comment (arttet) [inline cmd/shortener/routes.go:1]: It would be better if you move it in `internal`.

File: cmd/shortener/main.go [hunk 1]
```diff
@@ -18,29 +18,35 @@ func main() {
...
 }
 
-func registerRoutes(r *chi.Mux, storage storage.ShortenerStorage, cfg *config.Config) {
-	postShortenHandler := handler.NewPostShortenHandler(storage, cfg)
-	r.Handle("/", postShortenHandler)
-	r.Method(http.MethodPost, "/api/shorten", postShortenHandler)
-	r.Handle("/{id}", handler.NewGetURLHandler(storage, cfg))
-}
-
 func registerMiddleware(r *chi.Mux) {
```
Comment (arttet) [inline cmd/shortener/main.go:50]: It would be better if you move it in `cmd/shortener/routes.go`

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    restart: always
+    environment:
+      POSTGRES_PASSWORD: postgres
+    ports:
+      - 5432:5432
+    volumes:
+      - ./db-data:/var/lib/postgresql/data
+
+  adminer:
+    image: adminer
```
Comment (arttet) [inline docker/docker-compose.yml:16]: Please don't use it.

File: docker/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Use postgres/example user/password credentials
+version: '3.1'
+name: 'yandex-go-shortener'
+services:
+  db:
```
Comment (arttet) [inline docker/docker-compose.yml:5]: Where is your server?

File: internal/app/handler/get_ping.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"net/http"
+
+	"github.com/FoGezz/go-url-shortener/cmd/shortener/config"
+)
+
+type getPingHandler struct {
+	ShortenerHandler
+}
+
+func NewGetPingHandler(app *config.App) *getPingHandler {
```
Comment (arttet) [inline internal/app/handler/get_ping.go:15]: Why do you need it?

File: internal/app/handler/get_ping.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	return &getPingHandler{
+		ShortenerHandler: ShortenerHandler{
+			app,
+		},
+	}
+}
+
+func (h *getPingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if h.app.DBPool != nil {
+		err := h.app.DBPool.Ping(context.Background())
```
Comment (arttet) [inline internal/app/handler/get_ping.go:25]: You need to provide a context here

File: internal/app/storage/db_connection.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	"fmt"
+
+	"github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/postgres"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+//go:embed migrations/*.sql
+var migrationsDir embed.FS
```
Comment (arttet) [inline internal/app/storage/db_connection.go:16]: It would be better if it is in `migrations/*.sql`

File: internal/app/storage/shortener_db.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+func NewDBStorage(c *pgxpool.Conn) *DBStorage {
+	return &DBStorage{conn: c}
+}
+
+func (st *DBStorage) AddLink(full string, short string) {
+	_, _ = st.conn.Exec(context.Background(), "INSERT INTO links(long,short) VALUES ($1,$2);", full, short)
+}
+func (st *DBStorage) GetByShort(s string) (full string, found bool) {
+	err := st.conn.QueryRow(context.Background(), "SELECT long FROM links WHERE short = $1;", s).Scan(&full)
```
Comment (arttet) [inline internal/app/storage/shortener_db.go:21]: You have to provide a context for this function.

File: internal/app/storage/shortener_db.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+type DBStorage struct {
+	conn *pgxpool.Conn
+}
+
+func NewDBStorage(c *pgxpool.Conn) *DBStorage {
+	return &DBStorage{conn: c}
+}
+
+func (st *DBStorage) AddLink(full string, short string) {
+	_, _ = st.conn.Exec(context.Background(), "INSERT INTO links(long,short) VALUES ($1,$2);", full, short)
```
Comment (arttet) [inline internal/app/storage/shortener_db.go:18]: What if we would like to use another database manager system?

File: internal/app/storage/mock/storage.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (arttet) [inline internal/app/storage/mock/storage.go:1]: `internal/mocks/storage.go`

Comment (arttet) [reply internal/app/storage/mock/storage.go:1]: How did you generate it?

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/app/storage/shortener_db.go (hunk 8 above)
Comment (FoGezz) [reply internal/app/storage/shortener_db.go:18]: So we would better use squirrel :)

File: docker/docker-compose.yml (hunk 3 above)
Comment (FoGezz) [reply docker/docker-compose.yml]: added it as an "app" service

File: docker/docker-compose.yml (hunk 2 above)
Comment (FoGezz) [reply docker/docker-compose.yml]: ok

File: cmd/shortener/routes.go
Comment (FoGezz) [reply cmd/shortener/routes.go:1]: done

File: cmd/shortener/main.go (hunk 1 above)
Comment (FoGezz) [reply cmd/shortener/main.go:50]: done

File: internal/app/handler/get_ping.go (hunk 4 above)
Comment (FoGezz) [reply internal/app/handler/get_ping.go:15]: just have the same form as another handlers

File: internal/app/storage/db_connection.go (hunk 6 above)
Comment (FoGezz) [reply internal/app/storage/db_connection.go:16]: Не успеваю сделать это сегодня, что-то сразу не завелось у меня если из другого пакета импортировать (почему-то путь "migrations" вторым аргументом не нравится видимо), отправляю пока так

**Review by arttet — APPROVED** [review]

**Pull request merged**